File: services/ha-ai-agent-service/src/utils/performance_tracker.py
"""
Performance Tracking Utility

Tracks performance metrics for the chat endpoint and other operations.
Provides detailed timing information for optimization analysis.
"""

import logging
import time
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional
from collections import deque

logger = logging.getLogger(__name__)


@dataclass
class PerformanceMetric:
    """Single performance metric"""
    name: str
    start_time: float
    end_time: Optional[float] = None
    duration: Optional[float] = None
    metadata: Dict[str, Any] = field(default_factory=dict)


@dataclass
class PerformanceReport:
    """Performance report for an operation"""
    operation: str
    total_duration: float
    metrics: List[PerformanceMetric]
    timestamp: float


class PerformanceTracker:
    """Tracks performance metrics for operations"""
    
    def __init__(self, max_reports: int = 100):
        """
        Initialize performance tracker.
        
        Args:
            max_reports: Maximum number of reports to keep in memory
        """
        self.metrics: Dict[str, PerformanceMetric] = {}
        self.reports: deque = deque(maxlen=max_reports)
        self.max_reports = max_reports
    
    def start(self, name: str, metadata: Optional[Dict[str, Any]] = None) -> str:
        """
        Start tracking a metric.
        
        Args:
            name: Metric name
            metadata: Optional metadata
            
        Returns:
            Metric ID for ending the metric
        """
        metric_id = f"{name}_{int(time.time() * 1000)}_{id(self)}"
        self.metrics[metric_id] = PerformanceMetric(
            name=name,
            start_time=time.perf_counter(),
            metadata=metadata or {}
        )
        return metric_id
    
    def end(self, metric_id: str, metadata: Optional[Dict[str, Any]] = None) -> Optional[PerformanceMetric]:
        """
        End tracking a metric.
        
        Args:
            metric_id: Metric ID from start()
            metadata: Optional additional metadata
            
        Returns:
            Completed metric or None if not found
        """
        metric = self.metrics.get(metric_id)
        if not metric:
            logger.warning(f"Performance metric {metric_id} not found")
            return None
        
        end_time = time.perf_counter()
        duration = end_time - metric.start_time
        
        metric.end_time = end_time
        metric.duration = duration
        if metadata:
            metric.metadata.update(metadata)
        
        return metric
    
    def create_report(
        self, 
        operation: str, 
        metric_ids: List[str],
        additional_metadata: Optional[Dict[str, Any]] = None
    ) -> PerformanceReport:
        """
        Create a performance report for an operation.
        
        Args:
            operation: Operation name
            metric_ids: List of metric IDs to include
            additional_metadata: Optional additional metadata for the report
            
        Returns:
            Performance report
        """
        metrics = []
        for metric_id in metric_ids:
            metric = self.metrics.get(metric_id)
            if metric and metric.duration is not None:
                metrics.append(metric)
        
        total_duration = sum(m.duration for m in metrics if m.duration)
        
        report = PerformanceReport(
            operation=operation,
            total_duration=total_duration,
            metrics=metrics,
            timestamp=time.time()
        )
        
        # Add to reports
        self.reports.append(report)
        
        # Log report (convert seconds to milliseconds for display)
        total_duration_ms = total_duration * 1000
        logger.info(
            f"[Performance] {operation}: "
            f"Total: {total_duration_ms:.2f}ms ({total_duration:.3f}s), "
            f"Metrics: {len(metrics)}, "
            f"Details: {', '.join(f'{m.name}={m.duration*1000:.2f}ms' for m in metrics)}"
        )
        
        if additional_metadata:
            logger.debug(f"[Performance] {operation} metadata: {additional_metadata}")
        
        return report
    
    def get_reports(self) -> List[PerformanceReport]:
        """Get all reports"""
        return list(self.reports)
    
    def get_reports_for_operation(self, operation: str) -> List[PerformanceReport]:
        """Get reports for a specific operation"""
        return [r for r in self.reports if r.operation == operation]
    
    def get_average_duration(self, operation: str) -> float:
        """Get average duration for an operation"""
        reports = self.get_reports_for_operation(operation)
        if not reports:
            return 0.0
        total = sum(r.total_duration for r in reports)
        return total / len(reports)
    
    def clear(self) -> None:
        """Clear all metrics and reports"""
        self.metrics.clear()
        self.reports.clear()
    
    def export_reports(self) -> List[Dict[str, Any]]:
        """Export reports as dictionaries"""
        return [
            {
                "operation": r.operation,
                "total_duration": r.total_duration,
                "timestamp": r.timestamp,
                "metrics": [
                    {
                        "name": m.name,
                        "duration": m.duration,
                        "metadata": m.metadata
                    }
                    for m in r.metrics
                ]
            }
            for r in self.reports
        ]


# Global instance
_performance_tracker = PerformanceTracker()


def get_tracker() -> PerformanceTracker:
    """Get the global performance tracker instance"""
    return _performance_tracker


def start_tracking(name: str, metadata: Optional[Dict[str, Any]] = None) -> str:
    """Start tracking a metric (convenience function)"""
    return _performance_tracker.start(name, metadata)


def end_tracking(metric_id: str, metadata: Optional[Dict[str, Any]] = None) -> Optional[PerformanceMetric]:
    """End tracking a metric (convenience function)"""
    return _performance_tracker.end(metric_id, metadata)


def create_report(
    operation: str,
    metric_ids: List[str],
    additional_metadata: Optional[Dict[str, Any]] = None
) -> PerformanceReport:
    """Create a performance report (convenience function)"""
    return _performance_tracker.create_report(operation, metric_ids, additional_metadata)

File: services/ai-automation-ui/src/utils/performanceTracker.ts
/**
 * Performance Tracking Utility
 * 
 * Tracks performance metrics for the send button flow and other operations.
 * Provides detailed timing information for optimization analysis.
 */

export interface PerformanceMetric {
  name: string;
  startTime: number;
  endTime?: number;
  duration?: number;
  metadata?: Record<string, any>;
}

export interface PerformanceReport {
  operation: string;
  totalDuration: number;
  metrics: PerformanceMetric[];
  timestamp: number;
}

class PerformanceTracker {
  private metrics: Map<string, PerformanceMetric> = new Map();
  private reports: PerformanceReport[] = [];
  private maxReports = 100; // Keep last 100 reports

  /**
   * Start tracking a metric
   */
  start(name: string, metadata?: Record<string, any>): string {
    const id = `${name}_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
    this.metrics.set(id, {
      name,
      startTime: performance.now(),
      metadata,
    });
    return id;
  }

  /**
   * End tracking a metric
   */
  end(id: string, metadata?: Record<string, any>): PerformanceMetric | null {
    const metric = this.metrics.get(id);
    if (!metric) {
      console.warn(`Performance metric ${id} not found`);
      return null;
    }

    const endTime = performance.now();
    const duration = endTime - metric.startTime;

    const completedMetric: PerformanceMetric = {
      ...metric,
      endTime,
      duration,
      metadata: {
        ...metric.metadata,
        ...metadata,
      },
    };

    this.metrics.set(id, completedMetric);
    return completedMetric;
  }

  /**
   * Create a performance report for an operation
   */
  createReport(operation: string, metricIds: string[]): PerformanceReport {
    const metrics = metricIds
      .map((id) => this.metrics.get(id))
      .filter((m): m is PerformanceMetric => m !== undefined && m.duration !== undefined);

    const totalDuration = metrics.reduce((sum, m) => sum + (m.duration || 0), 0);

    const report: PerformanceReport = {
      operation,
      totalDuration,
      metrics,
      timestamp: Date.now(),
    };

    // Keep only last N reports
    this.reports.push(report);
    if (this.reports.length > this.maxReports) {
      this.reports.shift();
    }

    // Log to console for debugging
    console.log(`[Performance] ${operation}:`, {
      totalDuration: `${totalDuration.toFixed(2)}ms`,
      metrics: metrics.map((m) => ({
        name: m.name,
        duration: `${m.duration?.toFixed(2)}ms`,
        metadata: m.metadata,
      })),
    });

    return report;
  }

  /**
   * Get all reports
   */
  getReports(): PerformanceReport[] {
    return [...this.reports];
  }

  /**
   * Get reports for a specific operation
   */
  getReportsForOperation(operation: string): PerformanceReport[] {
    return this.reports.filter((r) => r.operation === operation);
  }

  /**
   * Get average duration for an operation
   */
  getAverageDuration(operation: string): number {
    const reports = this.getReportsForOperation(operation);
    if (reports.length === 0) return 0;
    const total = reports.reduce((sum, r) => sum + r.totalDuration, 0);
    return total / reports.length;
  }

  /**
   * Clear all metrics and reports
   */
  clear(): void {
    this.metrics.clear();
    this.reports = [];
  }

  /**
   * Export reports as JSON
   */
  exportReports(): string {
    return JSON.stringify(this.reports, null, 2);
  }
}

// Singleton instance
export const performanceTracker = new PerformanceTracker();

// Export convenience functions
export const startTracking = (name: string, metadata?: Record<string, any>) =>
  performanceTracker.start(name, metadata);

export const endTracking = (id: string, metadata?: Record<string, any>) =>
  performanceTracker.end(id, metadata);

export const createReport = (operation: string, metricIds: string[]) =>
  performanceTracker.createReport(operation, metricIds);

